index.js

```diff
@@ -194,4 +195,103 @@ export default async function pMap(
 	});
 }
 
+export function pMapIterable(
+	iterable,
+	mapper,
+	{
+		concurrency = Number.POSITIVE_INFINITY,
+	} = {},
+) {
+	if (iterable[Symbol.iterator] === undefined && iterable[Symbol.asyncIterator] === undefined) {
+		throw new TypeError(`Expected \`input\` to be either an \`Iterable\` or \`AsyncIterable\`, got (${typeof iterable})`);
+	}
+
+	if (typeof mapper !== 'function') {
+		throw new TypeError('Mapper function is required');
+	}
+
+	if (!((Number.isSafeInteger(concurrency) || concurrency === Number.POSITIVE_INFINITY) && concurrency >= 1)) {
+		throw new TypeError(`Expected \`concurrency\` to be an integer from 1 and up or \`Infinity\`, got \`${concurrency}\` (${typeof concurrency})`);
+	}
+
+	return {
+		[Symbol.asyncIterator]() {
+			const {next, nextError, complete, onCleanup, iterator} = createDeferredAsyncIterator();
+
+			const iterator_ = iterable[Symbol.iterator] === undefined ? iterable[Symbol.asyncIterator]() : iterable[Symbol.iterator]();
+
+			let currentIndex = 0;
+			let isFlushing = false;
+			let isComplete = false;
+
+			(async () => {
+				await onCleanup;
+				isComplete = true;
+			})();
+
+			const valuePromises = [];
+
+			async function flushValues() {
+				if (isFlushing) {
+					return;
+				}
+
+				isFlushing = true;
+
+				for await (const promise of valuePromises) {
+					try {
+						const value = await promise;
+
+						if (value === pMapSkip) {
+							continue;
+						}
+
+						next(value);
+					} catch (error) {
+						nextError(error);
+					}
+				}
+
+				isFlushing = false;
+			}
+
+			async function nextItem() {
+				if (isComplete) {
+					return;
+				}
+
+				const {done, value} = await iterator_.next();
+
+				if (isComplete) {
+					return;
+				}
+
+				const index = currentIndex;
+				currentIndex++;
+
+				if (done) {
+					isComplete = true;
+					return;
+				}
+
+				valuePromises.push(mapper(value, index));
+				flushValues();
+
+				await nextItem();
+			}
+
+			(async () => {
+				for (let index = 0; index < concurrency; index++) {
+					// eslint-disable-next-line no-await-in-loop
+					await nextItem();
+				}
+
+				complete();
+			})();
+
+			return iterator;
+		},
```

Comment:
Couldn't this simply be an async generator that yields the value wrapped in a concurrency limiter function. That way the whole logic is outside of the async iterator itself.

Something inspired by:
- https://github.com/sindresorhus/p-limit
- https://maximorlov.com/parallel-tasks-with-pure-javascript/
- https://stackoverflow.com/questions/40639432/what-is-the-best-way-to-limit-concurrency-when-using-es6s-promise-all

Reply:
Concurrency is now controlled by 2 options: `backpressure` and `concurrency`. Still possible to do this but may be harder to.